streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java

```diff
@@ -837,6 +837,49 @@
     KStream<K, V> through(final String topic,
                           final Produced<K, V> produced);
 
+    /**
+     * Materialize this stream to a auto-generated repartition topic and creates a new {@code KStream}
+     * from the auto-generated topic using default serializers, deserializers, producer's {@link DefaultPartitioner}.
+     * Number of partitions is inherited from the source topic.
+     *
+     * @return a {@code KStream} that contains the exact same (and potentially repartitioned) records as this {@code KStream}
+     * @see #repartition(Repartitioned)
+     * @see #repartition(KeyValueMapper, Repartitioned)
+     */
+    KStream<K, V> repartition();
+
+    /**
+     * Materialize this stream to a auto-generated repartition topic and creates a new {@code KStream}
+     * from the auto-generated topic using {@link Serde key serde}, {@link Serde value serde}, {@link StreamPartitioner},
+     * number of partitions and topic name part as defined by {@link Repartitioned}.
+     *
+     * @param repartitioned the {@link Repartitioned} instance used to specify {@link org.apache.kafka.common.serialization.Serdes},
+     *                      {@link StreamPartitioner} which determines how records are distributed among partitions of the topic,
+     *                      part of the topic name and number of partitions for a repartition topic, if repartitioning is required.
```

Comment:
yeah, that was something i wanted to verify actually. for example with DSL, use can do something like: `stream(...).mapValues().repartition()`. In this case repartition topic doesn't make much sense.

Reply:
I added integration to verify that no repartition topic is created if key-changing operation wasn't performed (or number of partitions was specified)

Reply:
If we don't repartition if user calls `repartition()` what is the purpose of the operation? The new operator is similar to `through()`, with the difference that Kafka Streams manages the topic.

Note, that one motivation for adding `repartition()` was, to allow users to repartition data before `transform()`. Atm, this in only possible via `through()` forcing users to create the corresponding topic manually what is cumbersome.

If a user does `stream(...).mapValues().repartition()` I agree that repartitioning is not really required, but I would see this as a user error. At the same time, I see the potential to address this in the optimization layer: if we detect this case, we could remove the `repartition()` operator. It seems to be a subtle difference, but it's semantically two different approaches IHMO -- what you suggest is to keep the operator but to make it a no-op, while I suggest to _remove_ the operator. This would result in the same topology but the code how it is achieve is different and I believe it's an important difference.

Reply:
I like the idea of addressing this on optimization layer. I think this depends on how end implementation would look like and what we gonna agree on in our main discussion thread on this PR. If we gonna have only `repartitioned` operator I guess it make sense to always force repartitioning. If we gonna go with `groupBy` and potentially `join` - optimization layer should be smarter about this in that case. I'll wait for an outcome of our discussion and update this accordingly.

Reply:
Based on comments from @vvcephei I've removed `repartition` operations from optimization logic altogether. Now, when calling `repartition` operations, corresponding repartition topic will be always created.